x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/create_upgradeable_rules_payload.ts

```diff
@@ -0,0 +1,190 @@
+import { type PrebuiltRuleAsset } from '../../model/rule_assets/prebuilt_rule_asset';
+import type { UpgradeableRule } from '../../model/rule_versions/get_version_buckets';
+import { getGroupNameForField, getMappedValueForField } from './diffable_rule_fields_mappings';
+import { assertRuleTypeMatch } from './assert_rule_type_match';
+import { createFieldUpgradeSpecifier } from './create_field_upgrade_specifier';
+import { determineFieldUpgradeStatus } from './determine_field_upgrade_status';
+import { FIELDS_PAYLOAD_BY_RULE_TYPE } from './create_fields_payload_by_rule_type';
+import { RULE_DEFAULTS } from '../../../rule_management/logic/detection_rules_client/mergers/apply_rule_defaults';
+import { calculateRuleFieldsDiff } from '../../logic/diff/calculation/calculate_rule_fields_diff';
+import { convertPrebuiltRuleAssetToRuleResponse } from '../../../rule_management/logic/detection_rules_client/converters/convert_prebuilt_rule_asset_to_rule_response';
+
+const RULE_DEFAULTS_FIELDS_SET = new Set(Object.keys(RULE_DEFAULTS));
+
+interface BuildUpgradePayloadProps {
+  upgradeableRules: UpgradeableRule[];
+  requestBody: PerformRuleUpgradeRequestBody;
+}
+
+export const createUpgradeableRulesPayload = ({
+  upgradeableRules,
+  requestBody,
+}: BuildUpgradePayloadProps) => {
+  const { pick_version: globalPickVersion = PickVersionValuesEnum.MERGED } = requestBody;
+  const ruleUpgradePayloads: PrebuiltRuleAsset[] = [];
+  const payloadErrors: Array<PromisePoolError<{ rule_id: string }>> = [];
+
+  upgradeableRules.forEach((upgradeableRule) => {
+    const targetRuleType = upgradeableRule.target.type;
+    const ruleId = upgradeableRule.target.rule_id;
+    const fieldsPayload = FIELDS_PAYLOAD_BY_RULE_TYPE.get(targetRuleType);
+
+    try {
+      if (fieldsPayload === undefined) {
+        throw new Error(`Unexpected rule type: ${targetRuleType}`);
+      }
+
+      assertRuleTypeMatch({ upgradeableRule, requestBody });
+
+      const calculatedRuleDiff = calculateRuleFieldsDiff({
+        base_version: upgradeableRule.base
+          ? convertRuleToDiffable(convertPrebuiltRuleAssetToRuleResponse(upgradeableRule.base))
+          : MissingVersion,
+        current_version: convertRuleToDiffable(upgradeableRule.current),
+        target_version: convertRuleToDiffable(
+          convertPrebuiltRuleAssetToRuleResponse(upgradeableRule.target)
+        ),
+      }) as AllFieldsDiff;
+
+      let rulePayload: Partial<PrebuiltRuleAsset> = {};
+
+      for (const field of fieldsPayload) {
+        rulePayload = {
+          ...rulePayload,
+          [field]: getValueForField({
+            field,
+            upgradeableRule,
+            globalPickVersion,
+            requestBody,
+            ruleId,
+            ruleFieldsDiff: calculatedRuleDiff,
+          }),
+        };
+      }
+
+      ruleUpgradePayloads.push(rulePayload as PrebuiltRuleAsset);
```

Comment:
Applied

Reply:
The problem is that the return value of this function is later passed to `upgradePrebuiltRules` and then that to our `DetectionRulesClient.upgradePrebuiltRule` which naturally takes a `PrebuiltRuleAsset` as type, not a `Partial<...>`.

It seems we need to do the casting some place, because we progressively build the PrebuiltRuleAsset. Or do you have in mind some other approach?